leetcode/src/19.c

```diff
@@ -0,0 +1,32 @@
+/**
+ * Definition for singly-linked list.
+ * struct ListNode {
+ *     int val;
+ *     struct ListNode *next;
+ * };
+ */
+
+struct ListNode *removeNthFromEnd(struct ListNode *head, int n) {
+    struct ListNode *p_free, *p = head;
+    int i, sz = 0;
+    while (p != NULL) {
+        p = p->next;
+        sz++;
+    }
+    if (n == sz) {
```

Comment:
What if we avooid this check with the following version of the code:
```
    struct ListNode sentry;
    sentry.next = head;

    struct ListNode *p_free, *p = &sentry;
    int i, sz = 0;

    while (p != NULL) {
        p = p->next;
        sz++;
    }

    p = &sentry;
    for (int i = 0; i < sz - n - 1; i++) {
        p = p->next;
    }
    
    struct ListNode *p_next_next = p->next->next;
    // we check only there if we need to free the resource.
    if (p_next_next == NULL) {
        free(p_next_next);
    }

    p->next = p_next_next;
    return sentry.next;
```
Would it be good? what do you think?

Reply:
I second this idea, the code is definitely cleaner this way. I should note that @alexpantyukhin's example code does contains some problems (duplicate `i` declaration, for example), but the idea's good

Reply:
I do not fully follow your suggestion, but if I understand correctly, you propose to introduce new entry node, so that there is no need to check if n == 1. After introducing entry node, I think the code should be as below. Need to call free() irrespective of value of n.

```
struct ListNode *removeNthFromEnd(struct ListNode *head, int n) {
    struct ListNode entry, *p_free, *p = head;
    entry.next = head;
    int sz = 0;
    while (p != NULL) {
        p = p->next;
        sz++;
    }
    p = &entry;
    for (int i = 0; i < sz - n; i++, p = p -> next)
    ;
    p_free = p->next;
    if (n != 1) {
        p->next = p->next->next;
    } else {
        p->next = NULL;
    }
    free(p_free);
    return entry.next;
}
```

Reply:
Pushed the commit for this change